client/client.go

```diff
@@ -1933,19 +1932,23 @@ func (c *Client) allocSync() {
 			}
 
 			var resp structs.GenericResponse
-			if err := c.RPC("Node.UpdateAlloc", &args, &resp); err != nil {
+			err := c.RPC("Node.UpdateAlloc", &args, &resp)
+			if err != nil {
+				// Error updating allocations, do *not* clear
+				// updates and retry after backoff
 				c.logger.Error("error updating allocations", "error", err)
-				syncTicker.Stop()
-				syncTicker = time.NewTicker(c.retryIntv(allocSyncRetryIntv))
-				staggered = true
-			} else {
-				updates = make(map[string]*structs.Allocation)
-				if staggered {
-					syncTicker.Stop()
-					syncTicker = time.NewTicker(allocSyncIntv)
-					staggered = false
-				}
+				syncTicker.Reset(c.retryIntv(allocSyncRetryIntv))
+				continue
 			}
+
+			// Successfully updated allocs, reset map and ticker.
+			// Always reset ticker to give loop time to receive
+			// alloc updates. If the RPC took the ticker interval
+			// we may call it in a tight loop before draining
+			// buffered updates.
+			updates = make(map[string]*structs.Allocation, len(updates))
```

Comment:
What's the reason for specifying `len(updates)` here? If the updates were successfully applied, shouldn't this reset to the same state as line 1908?

Reply:
It's a microoptimization to avoid map resizes. The length parameter to `make(map...)` gives a hint to the go runtime how many items the map is expected to hold. We don't know how many updates will happen, but the number we just performed is a reasonable heuristic. For lightly loaded nodes (dozens of mostly-long-lived allocations) this should never make a difference. For higher density or higher throughput nodes (eg hundreds of short lived batch allocs), this might be prevent some map growing. The risk is a using a tiny bit of extra memory when we don't necessarily need to (last update was large, next update is small). Likely hundreds of bytes or maybe single digit kb.